backend.go

```diff
@@ -27,6 +29,17 @@ var (
 	// when adding new alias name sources make sure to update the corresponding FieldSchema description in path_role.go
 	aliasNameSources          = []string{aliasNameSourceSAUid, aliasNameSourceSAName}
 	errInvalidAliasNameSource = fmt.Errorf(`invalid alias_name_source, must be one of: %s`, strings.Join(aliasNameSources, ", "))
+
+	// jwtReloadPeriod is the time period how often the in-memory copy of local
+	// service account token can be used, before reading it again from disk.
+	//
+	// Background for the selected value:
+	// Service account token expiration time is configurable by Kubernetes API
+	// server parameter service-account-max-token-expiration. The token is
+	// renewed when 80% of the expiration time is reached. The lowest allowed
```

Comment:
In an ideal world, I would prefer to set the TTL based on the `"exp"` claim in the token. However, it's complicated a little by the [`--service-account-extend-token-expiration`](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/#options) flag for kube-apiserver which extends actual lifetimes to a year, and just starts giving warnings when its "soft" expiry is hit, which defaults to 3607 seconds. The originally specified soft expiry is reported in a private claim for the token call `"warnafter"` and the body of the token looks something like this:

```json
{
  "aud": [
    "https://kubernetes.default.svc.cluster.local"
  ],
  "exp": 1669477045,
  "iat": 1637941045,
  "iss": "https://kubernetes.default.svc.cluster.local",
  "kubernetes.io": {
    "namespace": "default",
    "pod": {
      "name": "nginx",
      "uid": "89e8c5b0-a2ff-44b4-9743-f11c477eac46"
    },
    "serviceaccount": {
      "name": "default",
      "uid": "d41af982-c2f6-416f-8181-793aae2a31b4"
    },
    "warnafter": 1637944652
  },
  "nbf": 1637941045,
  "sub": "system:serviceaccount:default:default"
}
```

WDYT of respecting the "warnafter" claim if it exists (to stop us triggering warnings), and if not setting the TTL based on "exp"? Although not the simplest, it seems the least complicated robust solution to me. And given the token is tied to our own pod + the service account, I think the only way that could break is if someone deleted the service account, and that's probably not worth defending against anyway :)

Reply:
Some notes about possible complications related reload logic based on expiration time from claims:

* The code should support following alternatives: (1) legacy token without `exp` and `warnafter`, (2) new token with `exp` and `warnafter` and (3) token with only `exp`.
* Reloading should be done _before_ the token expires: `exp - [offset]`. This is to avoid potential authentication failure at the very end of the token validity period.
* Kubelet will injected the new token some time in advance (it calculates some % of configured token expiration time). There is a possibility where we end up reloading too early and end up with the same token. The reload logic should avoid any kind of "busy-looping" if the newly reloaded token still is considered "old" by the implemented criteria.

Reply:
I am happy with once per minute too, especially if there is a recommendation and prior art for that in client-go.

For posterity, my suggestion would have been to consider the token stale once it reached 80% or 90%. If the token hadn't been refreshed by k8s, then even a fresh read of it would be considered stale. Then we could continue attempting TokenReviews with "stale" tokens, but we would also re-read the token every time until we got a refreshed one. Not really relevant anymore though, let's go with once per minute :)

> I'm bit confused here, since client would not use the shorter-lived tokens that are produced by service account volume projection.

Not sure if I understand correctly, but is the point here that the default admission injected token does not have a configurable TTL? i.e. at `/var/run/secrets/kubernetes.io/serviceaccount/token`? While true, that default-mounted token can be overridden, and I think some people will _want_ to override it because of the --service-account-extend-token-expiration flag which can't be turned off for managed solutions like GKE. So I'd like to make sure it still works for those users, and while we could add a configuration option, it would be better if we can make it Just Work™️ without any configuration.

Reply:
Cool, I've changed the refresh period to 1 minute to match the recommendation!

My confusion was just about the ranges:

* Kubernetes allows setting the minimum period to 1 hour for `/var/run/secrets/kubernetes.io/serviceaccount/token`.
* Kubernetes allows setting the minimum period to 10 minute for custom tokens produced by service account volume projection.

The first token is used to talk to Kubernetes API server, so why would one recommend 1 minute refresh period for a token that is refreshed every 1 hour at most?  This does not matter in practice though!  I'm also extremely happy with Just Works (tm) :-)